lib/src/neu/generators/openapi3/openapi3-type-util.ts

```diff
@@ -0,0 +1,315 @@
+          values: nonNullTypes.map(t => t.value),
+          nullable
+        });
+      } else if (areStringLiteralTypes(nonNullTypes)) {
+        return stringSchema({
+          values: nonNullTypes.map(t => t.value),
+          nullable
+        });
+      } else if (areFloatLiteralTypes(nonNullTypes)) {
+        return numberSchema({
+          values: nonNullTypes.map(t => t.value),
+          format: "float",
+          nullable
+        });
+      } else if (areIntLiteralTypes(nonNullTypes)) {
+        return integerSchema({
+          values: nonNullTypes.map(t => t.value),
+          format: "int32",
+          nullable
+        });
+      } else {
+        return {
+          nullable: nullable || undefined,
+          oneOf: nonNullTypes.map(t =>
+            typeToSchemaOrReferenceObject(t, typeTable)
+          ),
+          discriminator: unionTypeToDiscrimintorObject(type, typeTable)
+        };
+      }
+  }
+}
+
+function unionTypeToDiscrimintorObject(
+  unionType: UnionType,
+  typeTable: TypeTable
+): DiscriminatorObject | undefined {
+  if (unionType.discriminator === undefined) return undefined;
+
+  const nonNullTypes = unionType.types.filter(isNotNullType);
+
+  // Discriminator mapping is only supported for reference types
+  const objectReferenceOnly = nonNullTypes.every(
+    t => isReferenceType(t) && isObjectType(dereferenceType(t, typeTable))
+  );
+  if (!objectReferenceOnly) return { propertyName: unionType.discriminator };
+
+  const mapping = nonNullTypes.reduce<{ [key: string]: string }>((acc, t) => {
+    // Sanity check and type cast
+    if (!isReferenceType(t)) {
+      throw new Error("Unexpected error: expected reference type");
+    }
+
+    const concreteType = dereferenceType(t, typeTable);
+
+    // Sanity check and type cast
+    if (!isObjectType(concreteType)) {
+      throw new Error("Unexpected error: expected object reference type");
+    }
+
+    // Retrieve the discriminator property
+    // Discriminator properties cannot be optional - we assume this is handled by the type parser
+    const discriminatorProp = concreteType.properties.find(
+      p => p.name === unionType.discriminator
+    );
+    if (discriminatorProp === undefined) {
+      throw new Error(
+        "Unexpected error: could not find expected discriminator property"
+      );
+    }
+
+    // Extract the property type - this is expected to be a string literal
+    const discriminatorPropType = dereferenceType(
+      discriminatorProp.type,
+      typeTable
+    );
+    if (!isStringLiteralType(discriminatorPropType)) {
+      throw new Error(
+        "Unexpected error: expected discriminator property type to be a string literal"
+      );
+    }
+
+    acc[discriminatorPropType.value] = referenceObjectValue(t.name);
+    return acc;
+  }, {});
+
+  return {
+    propertyName: unionType.discriminator,
+    mapping
+  };
+}
+
+function referenceTypeToSchema(
+  type: ReferenceType,
+  nullable?: boolean
```

Comment:
I would do 
```suggestion
  nullable = false
```

Reply:
as above - Where possible throughtout the OpenAPI spec, I've avoided including any details are aren't required. It would be preferable if this maintained that behaviour.